benches/aarch64/mod.rs

```diff
@@ -0,0 +1,48 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+extern crate criterion;
+extern crate linux_loader;
+extern crate vm_memory;
+
+use linux_loader::configurator::fdt::FdtBootConfigurator;
+use linux_loader::configurator::BootConfigurator;
+use linux_loader::configurator::BootParams;
+use vm_memory::{ByteValued, GuestAddress, GuestMemoryMmap};
+
+use criterion::{black_box, Criterion};
+
+const MEM_SIZE: usize = 0x100_0000;
+const FDT_MAX_SIZE: usize = 0x20;
+
+fn create_guest_memory() -> GuestMemoryMmap {
+    GuestMemoryMmap::from_ranges(&[(GuestAddress(0x0), MEM_SIZE)]).unwrap()
+}
+
+#[derive(Clone, Copy, Default)]
+pub struct FdtPlaceholder([u8; FDT_MAX_SIZE]);
+
+unsafe impl ByteValued for FdtPlaceholder {}
+
+fn build_fdt_boot_params() -> BootParams {
+    let fdt = FdtPlaceholder([0u8; FDT_MAX_SIZE]);
+    let fdt_addr = GuestAddress((MEM_SIZE - FDT_MAX_SIZE - 1) as u64);
+    BootParams::new::<FdtPlaceholder>(&fdt, fdt_addr)
+}
+
+pub fn criterion_benchmark(c: &mut Criterion) {
+    let guest_mem = create_guest_memory();
+    let fdt_boot_params = build_fdt_boot_params();
+    c.bench_function("configure_fdt", |b| {
+        b.iter(|| {
+            black_box(FdtBootConfigurator::write_bootparams::<GuestMemoryMmap>(
+                fdt_boot_params.clone(),
```

Comment:
With this approach it looks like we're also including the time it takes to clone the fdt_boot_params into consideration. Do we have information of how expensive clone might be in this scenario? I am wondering if we can re-write it somehow such that clone is not executed in the benchmark function.

Reply:
Indeed, other artefacts made their way into the bench loop (the `clone`s, a `Cursor::new`). I couldn't find a way to avoid them. I'll check how long they take.

Reply:
Checked and:
* `Cursor::new` takes a meager ~1% of the time it takes to `load` the kernel
* but the `clone` takes 50% of the total `write_bootparams` time

I can't think of a way to eliminate the `clone` other than refactoring `write_bootparams` to take a `BootParams&` instead, if that's possible. Thoughts?

Reply:
Wasn't that hard to pass the `BootParams` by ref, and the benchmark numbers are looking better :+1: Nice catch!